### ISSUE TITLE
Custom URI scheme for receiving addresses

### ISSUE THREAD

**Issue by GeorgeFlerovsky** [uniVocity/free-commerce#4]
It would be really cool if the various cardano wallets developed a standard for a custom URI scheme to be used to open receiving addresses in wallets. Something like: [ada://addr1qyluakavdhfmd54s8...](ada://addr1qyluakavdhfmd54s8...) or maybe [crypto://...](crypto://...) or [crypto-pay://...](crypto-pay://...)

This would be complementary to the existing ways to reference an address: plain text and QR codes. I think it would something that most shopkeepers would want to include in the generated email that gets sent to customers for orders. 

QR codes are really only convenient when you see the QR code on one device (e.g. desktop computer, paper poster, etc) and scan it with another device (e.g your phone). The scanning device requires a camera, and it's quite awkward to scan a QR code from the same device as where you see the QR code.

I'm not entirely sure how to make a new custom URI scheme, but I think this is something worth exploring with the Cardano wallets and this e-commerce project.

**Comment by jbax** [comment]
Yes! Other than payments, one thing I want to do eventually is to create
one click actions that can also be supported by the scheme.

This will allow stores to generate a url which instructs the wallet to
immediately perform an action.

Actions encoded in a QR code can be printed and added to store windows,
boxes, packages, bags, video ads, t-shirts, trucks, you name it.

So other than payments, we can also encode:

 * checkout: sellers could generate a url/qr code to sell a toy and add a
"get this toy too, just scan the qr code below"

 * discounts: same as above but to apply a discount code before completing
the purchase.

 * more info: load more details about an item,or take user to the purchase
receipt, or to the transaction details in the blockchain.

I've been thinking about it for a while and it will be supported by this
wallet eventually.

**Comment by GeorgeFlerovsky** [comment]
Awesome! Sounds ambitious.

One thing to consider RE QR printouts. You intended to generate one address per order, right? That's hard to accomodate with a paper printed QR code in a physical store. On the other hand, if you go with one address per shop item, you compromise a bit on privacy – anyone can check what the store received for this item.

As for the rest, I think there should be a clear line drawn between store logic and wallet/payment system logic. For example, when I buy something on Amazon using Paypal to pay: Amazon handles all the discounts/promotions/upselling, while Paypal only has to deal with making the payment go through in the required currency. Do you see your "free commerce" plugin as a "paypal" or do you intend to draw the lines further outward?

**Comment by GeorgeFlerovsky** [comment]
Or, by "QR printout", do you mean a store owner has to print out an individual QR code for every single chocolate bar she sells? ;)

**Comment by jbax** [comment]
Payments are always unique, the associated qr code can only be used once.

A qr code checkout for example initiates the process of checking out an
item (payment comes after). Until the order is confirmed - still before
payment - discounts can be applied as well.

The idea is that the wallet will communicate with the free commerce server
- which enables the purchase processing. The wallet itself will initially
only be used to pay the final order amount.

To the users who are buying using a "buy this item too", it will look like
they're buying from the wallet without even seeing shopify or the online
store.

It works like this:
- scan qr code with checkout details.
- wallet opens with a "confirm your purchase".
- user confirms it's the product they want, and the shipping details
(address and other details could be saved in the wallet by a plug-in)
- server associates confirmed order to a payment address, and waits for
payment.
- order is paid by the wallet.

Done. Lots of things happen under the hood but the consumer just clicks to
buy then confirms.

Hope this clarifies.

**Comment by GeorgeFlerovsky** [comment]
Oh, I see. So the QR printout doesnt encode a receiving address, but rather a link to initiate a checkout action, which triggers the wallet. Yup, that works.

**Comment by GeorgeFlerovsky** [comment]
I imagine that triggering a wallet to open with a "confirm your purchase" would probably require a custom URI scheme link, at some point. Right?

**Comment by jbax** [comment]
Yes. The URI scheme is what enables everything.
